### ISSUE TITLE
Improve GraphQL TypedDocumentNode usage in the new MSW.js version

### ISSUE THREAD

**Issue by MatejBransky** [kettanaito/msw#1881]
### Scope

Improves an existing behavior

### Compatibility

- [ ] This is a breaking change

### Feature description

Hi, we have successfully used MSW.js v1 with generated GraphQL typed documents (see type notation in screenshots) so we have to just import the TypedDocumentNode document itself and the rest has been inferred. It was one of the selling points in our team.

<img width="806" alt="image" src="https://github.com/mswjs/msw/assets/20944757/8e23f3e1-ee4b-4dbc-8417-6b161b1c4c47">

But this behavior doesn't work in the latest version because we have to return type-agnostic `HttpResponse.json()` which can't know what the response type should be.

<img width="408" alt="image" src="https://github.com/mswjs/msw/assets/20944757/22b7a89e-7376-4e4c-86b2-9bff3aaa56a8">

Am I missing some new way to do it with the same comfort, or is this completely gone?

**Comment by MatejBransky** [comment]
The workaround is to pass the query type explicitly. But DX is worse than the previous major version.

<img width="1016" alt="image" src="https://github.com/mswjs/msw/assets/20944757/db8f9c33-5ed6-49c1-8db8-0ed0c7ef2c52">

**Comment by kettanaito** [comment]
Hey. This looks like a bug with the body type inference. Thanks for sharing this! A type test is clearly missing on our part.

**Comment by MatejBransky** [comment]
Glad to hear that it is just a bug. Thank you for your work! You guys rock! 🚀

**Comment by pleunv** [comment]
@MatejBransky I can't reproduce this behavior myself, and I have working type inference for the `HttpResponse.json({...})` object by simply passing the `TypedDocumentNode` document. Do you still encounter this today?

One thing worth noting is that if you're using `__typename`, you need an `as const` annotation or inference will break:

```ts
graphql.query(GetUser, () => 
  HttpResponse.json({
    __typename: 'User' as const,
   // ... user properties, with inference
  })
);
```

**Comment by aaronadamsCA** [comment]
@pleunv Could you please confirm that if you manually pass the query type as a type parameter, everything works as you expect it:

```ts
graphql.query<GetUserQuery>(GetUser, () => 
  HttpResponse.json({
    __typename: 'User',
   // ... user properties, with inference
  })
);
```

If so, that more or less confirms the same problem I'm seeing: when I call `graphql.query(typedDocumentNode, resolver)`, TypeScript seems to infer the type of `Query` from `resolver` instead of `typedDocumentNode`. This would explain both the type mismatch and why `as const` resolves the problem.

Unfortunately my TypeScript skills are right in that sweet spot where I am pretty sure I understand the problem, but I have absolutely no idea how to solve it.

**Comment by aaronadamsCA** [comment]
**Edit:** I tried the below locally using `pnpm patch` and it didn't work, but maybe it helps point toward a solution.

<details>

Looking at this:

https://github.com/mswjs/msw/blob/e56f6243bffc3237319f50b63ac0e38b2f4a9c73/src/core/graphql.ts#L26-L36

And given this: https://github.com/microsoft/TypeScript/issues/14829#issuecomment-320754731

> Just an update - `T & { } ` creates a "lower-priority" inference site for `T` _by design_. I would move this from the "definitely don't depend on this" column to the "it's probably going to work for the foreseeable future" column.

The fix might be as simple as this:

```ts
export type GraphQLRequestHandler = <
  Query extends GraphQLQuery = GraphQLQuery,
  Variables extends GraphQLVariables = GraphQLVariables,
>(
  operationName:
    | GraphQLHandlerNameSelector
    | DocumentNode
    | TypedDocumentNode<Query, Variables>,
  resolver: GraphQLResponseResolver<Query & {}, Variables & {}>,
  options?: RequestHandlerOptions,
) => GraphQLHandler
```

Which should hopefully encourage TypeScript to infer `Query` and `Variables` from `operationName` instead of `resolver`.

</details>

**Comment by pleunv** [comment]
> @pleunv Could you please confirm that if you manually pass the query type as a type parameter, everything works as you expect it:
> 
> ```ts
> graphql.query<GetUserQuery>(GetUser, () => 
>   HttpResponse.json({
>     __typename: 'User',
>    // ... user properties, with inference
>   })
> );
> ```

Can confirm that works as well, yeah. You have two options when using `TypedDocumentNode`:

- Importing the operation data & variable types and passing them as generic type annotations to `graphql.query<TData, TVars>()` and `graphql.mutation<TData, TVars>()`
- Marking all string literals as `as const`

Both are a bit annoying, I went with the latter for now :)

**Comment by kettanaito** [comment]
@aaronadamsCA, thanks for the suggestion! I don't think the `T & {}` works how you expect. It just creates a unique type, preventing type matching for broader types. A small example to illustrate:

```ts
function accept(name: 'john' | string): void {}
// All string inputs will be "coerces" to the shared
// type—which is just string. You won't get autocompletion
// on the "john" type of the "name".
accept('john')
accept('sarah')

function accept(name: 'john' | string & {}): void {}

// By using "& {}", we tell TS to differentiate between
// the "john" base type (string) and the "any other string"
// type. As a result, we get a proper autocompletion for "john":
accept('john')

// We also get any strings as "name" supported.
accept('sarah')
```

It doesn't really affect the order in which the generic parameters are inferred, afafik.

**Comment by kettanaito** [comment]
I've added a type test for `TypedDocumentNode` and it passes without issues: https://github.com/mswjs/msw/pull/2018. 

Can it be that the `TypedDocumentNode` type itself has changed? We don't depend on GraphQL Code Generator in MSW, instead we are just replicating the `TypedDocumentNode` type to provide query and variable types inference:

https://github.com/mswjs/msw/blob/e56f6243bffc3237319f50b63ac0e38b2f4a9c73/src/core/graphql.ts#L17-L24

This means if the GQL Code Gen was to change this type, like swapping the order of generics or changing the return type, MSW would stop inferring from it. 

This is also why the [suggestion](https://github.com/mswjs/msw/issues/1881#issuecomment-1929477973) from @aaronadamsCA may work. If the `TypedDocumentNode` has the `__typename` in its return type, that will trigger the type inference. Now, trying to find where the `TypedDocumentNode` type is defined...

**Comment by kettanaito** [comment]
Alas, just checked the `TypedDocumentNode` type and it remains the same: https://github.com/dotansimha/graphql-typed-document-node/blob/925c2b0483ae602eec38253acf602c2aefe383ec/packages/core/src/index.ts#L3-L10.

I've also added `@graphql-typed-document-node/core` to use in GraphQL type testing with `TypedDocumentNode`, and the inference still works as of MSW 2.1.6.  

## Needs help

Does anybody have a reproduction repository for this? Without it, I would have to close this issue as non-reproducible.

**Comment by aaronadamsCA** [comment]
I tried the TypeScript `v5.4.0-beta` which ships a new `NoInfer` utility type. That didn't solve the problem either, so I may have been off the mark RE: inference. It might be a contravariance problem or something? The type error isn't quite what I'd expect it to be.

@kettanaito, I would expect #2018 to pass because your typed document node expected `name` to be a `string`. Try this:

```diff
 const GetUserQuery = {} as TypedDocumentNode<
   {
     user: {
-      name: string
+      name: 'John'
     }
   },
   { userId: string }
```

If that doesn't reproduce, I'll see what I can put together (I tried earlier today but the CodeSandbox starter seems to be broken, the dependencies won't install).

**Comment by kettanaito** [comment]
@aaronadamsCA, I've just tried a narrower type for `user.name` and the test still passes. The point there is that the inference was correct, so no matter what types you provide, MSW will respect hose.

I think the only case when that won't happen is if the type of `TypedDocumentNode` (the actual liter value) doesn't extend this type MSW expects (which is also the type that GQL Code Gen generates):

https://github.com/mswjs/msw/blob/e56f6243bffc3237319f50b63ac0e38b2f4a9c73/src/core/graphql.ts#L17-L24

Can you please look at the type of your document node and post it here? What fields does it have? There must be some divergence that causes the inference to break. 

Once again, a reproduction repository would be best.

**Comment by MatejBransky** [comment]
I've created this repo: https://github.com/MatejBransky/graphql-example

Try to complete [the already existing mocked query handler](https://github.com/MatejBransky/graphql-example/blob/main/tests/msw/handlers.ts#L10-L12) in the file `tests/msw/handlers.ts`.

The "foobar" doesn't trigger any TS error and expected "state" field should be suggested but it's not.

The suggested shape should be:
```ts
HttpResponse.json({
  data: {
    info: {
      state: 'something'
    }
  }
})
```

**Comment by aaronadamsCA** [comment]
And here's another repro:

https://stackblitz.com/edit/typescript-hr75ym?file=reproduction.ts

**Comment by kettanaito** [comment]
@MatejBransky, thanks for providing the repo! If you open `src/generated/graphql.ts`, you can spot a TypeScript error right here:

```ts
import { TypedDocumentNode as DocumentNode } from '@graphql-typed-document-node/core'
```

There are no type definitions for the `@graphql-typed-document-node/core` so `DocumentNode` will be `any`. Once you install that package, you get the right type inference from MSW:

```sh
pnpm i @graphql-typed-document-node/core
```
```
...
Type '{ data: { foobar: string; }; }' is not assignable to type 'GraphQLResponseBody<GetAppInfoQuery>'.
        Types of property 'data' are incompatible.
          Property 'info' is missing in type '{ foobar: string; }' but required in type 'GetAppInfoQuery'.ts(2345)
```

It reports `foobar` as unknown and also autocompletes the `info` key as expected.

**Comment by kettanaito** [comment]
@aaronadamsCA, I'm getting what looks like correct type validation in that sandbox as-is:

<img width="710" alt="Screenshot 2024-02-06 at 16 52 33" src="https://github.com/mswjs/msw/assets/14984911/50d46edf-9074-42fb-b8ae-c610d74d937c">

Is the expected state different from this?

**Comment by MatejBransky** [comment]
@kettanaito Awesome! Thank you! I can close it I guess?

**Comment by kettanaito** [comment]
@MatejBransky, glad it helped!

Give me a moment, I think I finally understand the problem with the sandbox from @aaronadamsCA. Investigating...

**Comment by kettanaito** [comment]
@aaronadamsCA, that looks like an issue but I can't figure out where it belongs. Sharing my thoughts. 

The `HttpResponse.json()` body generic parameter will infer the type from the given value _as broadly as possible_. So `foo: 'BAR'` gets inferred as `foo: string`. MSW still correctly infers the return type from the query because it's that type that causes the error:

- The resolver is annotated to return `foo: 'BAR'` type while the actual response is `foo: string` (a more permissive type). 

I can see now what you were talking about. This may be beyond my skill. Perhaps @andarist can have a piece of advice here once he's got a minute. Here's a [playground](https://www.typescriptlang.org/play?ts=5.3.3#code/PTAECEEMGcEsGNIBskE8A0oAWB7A7qAC45ECmKoAKqgA6kDK8ATrDYUSbAHYBmpTVAFAhQMUACoukJk3xpxo6KBo5ocAEZJSAOkE8Arl3iFYOLqABW0MwB5KAPgAU6nABNUALioBKLwDccWFdQAG8AX0FhMABRAA86Y1JXL0pQAF5Q0S8AcnAAQQAlbNAIkTzjfWQU9MzIL2hCFi4AcxLBKzNHEKzQXMLisO9BIA) that distills this example to the minimum.

**Comment by kettanaito** [comment]
I will close this issue because it needs no changes on MSW's side. What @andreawyss is experiencing is not related to `TypedDocumentNode`. You can reproduce that even with a response body generic using the `http` handler. I will keep an eye on this thread if any suggestions come our way on how to force TypeScript to infer the type as narrowly as possible.

**Comment by MatejBransky** [comment]
Ok, I've found one issue even after adding the missing typings - if I don't fill the expected response the TS doesn't complain about the missing type. 🤔 

<img width="625" alt="image" src="https://github.com/mswjs/msw/assets/20944757/855dba6e-1c9a-406c-8cc6-20dc0b75f6fc">


And nested fields are not suggested. 👀 
<img width="432" alt="image" src="https://github.com/mswjs/msw/assets/20944757/edd0c5e4-28bd-44fb-a3b8-acb2b3865a41">

This wasn't the issue in the v1:
<img width="464" alt="image" src="https://github.com/mswjs/msw/assets/20944757/61f32a45-8a34-4e06-b610-b556b479e138">


I've pushed this example to the mentioned repo.

**Comment by aaronadamsCA** [comment]
> @aaronadamsCA, that looks like an issue but I can't figure out where it belongs. Sharing my thoughts.
> 
> The `HttpResponse.json()` body generic parameter will infer the type from the given value _as broadly as possible_. So `foo: 'BAR'` gets inferred as `foo: string`. MSW still correctly infers the return type from the query because it's that type that causes the error:
> 
> * The resolver is annotated to return `foo: 'BAR'` type while the actual response is `foo: string` (a more permissive type).

Yep, you've got it, @kettanaito. Therefore it happens with every `__typename` and with every `enum` in the schema.

> I can see now what you were talking about. This may be beyond my skill. Perhaps @Andarist can have a piece of advice here once he's got a minute. Here's a [playground](https://www.typescriptlang.org/play?ts=5.3.3#code/PTAECEEMGcEsGNIBskE8A0oAWB7A7qAC45ECmKoAKqgA6kDK8ATrDYUSbAHYBmpTVAFAhQMUACoukJk3xpxo6KBo5ocAEZJSAOkE8Arl3iFYOLqABW0MwB5KAPgAU6nABNUALioBKLwDccWFdQAG8AX0FhMABRAA86Y1JXL0pQAF5Q0S8AcnAAQQAlbNAIkTzjfWQU9MzIL2hCFi4AcxLBKzNHEKzQXMLisO9BIA) that distills this example to the minimum.

I'm not entirely sure the solution involves preventing type widening (which I don't think is possible?), but rather in controlling type inference. There must be a way to tell TypeScript to infer the query type _from_ `operationName` and then apply it _to_ `resolver`. I just have no idea how.

**Comment by kettanaito** [comment]
@aaronadamsCA, to clarify, this part is working as you describe:

> There must be a way to tell TypeScript to infer the query type from operationName and then apply it to resolver

It infers your query and variable types and annotates the resolver's return type to match the query type. There are no issues with this.

The issue is the inference here:

```ts
HttpResponse.json<T>(body: T)
```

What `HttpResponse.json()` returns must be compatible with what the response return type has inferred. It is during the json body inference that the given type is inferred as broadly as possible. 

I don't believe there's a way for the `HttpResponse.json()` to infer type _from_ the resolver surrounding it. There may be TS magic to do this but I don't posses that incantation.

**Comment by aaronadamsCA** [comment]
I've created a new reproduction that's as short as possible:

https://stackblitz.com/edit/msw-response-type-mismatch?file=index.ts

I think this is the best illustration that there really is a bug here, because it can be worked around by explicitly passing the same type parameter that is already being inferred. You can see that in both cases the type parameters to `graphql.query` are the same, yet one works and one doesn't.

I don't know, maybe it's a TypeScript issue.

**Edit:** Ah, simultaneous posts. I understand what you're saying, but then I would expect the workaround to fail as well. You can see that the inferred type of `HttpResponse.json()` differs between `badHandler` and `goodHandler`—for no apparent reason, really.

**Comment by kettanaito** [comment]
@aaronadamsCA, may I please ask you to open a new issue about this? I would like to take a proper look at it. It's just not related to `TypedDocumentNode`, really. Thanks.

**Comment by MatejBransky** [comment]
So just to clarify - does that mean that in MSW v1 was possible to infer the return type from the typed document node immediately without any `__typename: 'typename' as const` and in v2 you have to add that field in order to get the same type hint?

**Comment by Andarist** [comment]
The easiest way to enforce `const`-like inference is to use `const` type parameters that were introduced in TS 5.0:
[TS playground](https://www.typescriptlang.org/play?ts=5.3.3#code/PTAECEEMGcEsGNIBskE8A0oAWB7A7qAC45ECmKoAKqgA6kDK8ATrDYUSbAHYBmpTVAFAhQMUACoukJk3xpxo6KBo5ocAEZJSAOkE8Arl3iFYOLqABW0MwB5KAPgAU6nABNUALioBKLwDccWFdQAG8AX0FhMABRAA86Y1JXL0pQAF5Q0S8AcnAAQQAlbNAIkTzjfWQU9MzIL2hCFi4AcxLBKzNHEKzQACJ8gt6S7wBuKNAAPQB+SJFYdiZSAEd9WEWlSnpQAFZtAAY9Q2NTc3gzBoApay4bM64GqicXdxTfUACg0Ii7y+uunv6hSGYVG42mszAhCwsCUrhwpGgXGyC3IqFAZioW12e1A6n07Hg+hkpC4hDQoHm6K45OYpEghARoB+7AaTWa0HGUNITLMrnmJ2QRFo3MQ5nU3NIsQZXFcSWw-G5xFA0H0NBUTHYAFscIshXQlEhYABrblcfSa8VMJSkQjwQSEYWgAAipFINAAwudCHZ7DUQoJQKAANoAaQp5hNqBwPCoAF0UqHY6BJdLXEpWdxWlMqInQF4XW7PfdvQ4xmExgYjCYMYs-PxoEkrrYHM43J5na6PV6fW8PsFwoJa-XG39unU+gNgaCROCgA)

**Comment by MatejBransky** [comment]
Ok, so the workaround/solution is [in the comment of the related PR](https://github.com/mswjs/msw/issues/2023#issuecomment-1943876229). Thanks @pleunv ! 🙏

**Comment by kettanaito** [comment]
## Released: v2.11.2 🎉

This has been released in v2.11.2!

- 📄 [**Release notes**](https://github.com/mswjs/msw/releases/tag/v2.11.2)
- 📦 [npm package](https://www.npmjs.com/package/msw/v/2.11.2)

Make sure to always update to the latest version (`npm i msw@latest`) to get the newest features and bug fixes.

---

_Predictable release automation by [@ossjs/release](https://github.com/ossjs/release)_.
